src/main/scala/strawman/collection/immutable/Spandex.scala

```diff
@@ -0,0 +1,247 @@
+  * <li>amortised constant time <code>prepend</code> and <code>append</code> (depending on the complexity of <code>scala.Array.copy</code>)</li>
+  * <li>efficient indexed access</li>
+  * <li>linear time iteration (i.e. <code>iterator</code> and <code>foreach</code>) with low constant factors
+  * <li>reasonable memory usage (at most double (plus constants) that of an array with the same number of elements)
+  * </ul>
+  * <br/>
+  * The underlying array is only mutated on <code>prepend</code> and <code>append</code> but never more than once for any given position (any later modification attempts to an already modified position results in a copy being made of the underlying array). To guarantee that only a single thread can write to an array slot synchronization is used during slot assignment in <code>prepend</code> and <code>append</code>. All other access is non synchronised.
+  * <br/>
+  * <br/>
+  * Expansion occurs when the underlying array is full on the effected side; the new array will be populated to have its center position adjusted by the unused capacity (margin) on the non effected side according to the following formula: <code>starting position = (length + margin) / 2</code> (i.e. for a six element spandex with two slots unused at the end <code>[a, b, c, d, e, f, , ]</code>, a prepend operation would make the expanded array have a length of 12 and a starting position of <code>(6 + 2) / 2 = 4</code>, <code>[ , , , , a, b, c, d, e, f, , ]</code>). This expansion scheme leads to more free slots being allocated on the side mostly expanded (a margin of zero will allocate an equal amount of free slots on both sides).
+  */
+sealed abstract class Spandex[+A]
+  extends Seq[A]
+    with SeqLike[A, Spandex]
+    with LinearSeq[A]
+    with MonoBuildable[A, Spandex[A]]
+    with PolyBuildable[A, Spandex] {
+  protected def index: Int
+  protected def array: scala.Array[Any]
+  protected def primary: Spandex.Primary[A]
+
+  override def size: Int = length
+  override def knownSize: Int = length
+
+  protected def element(i: Int): A = scala.runtime.ScalaRunTime.array_apply(array, i).asInstanceOf[A]
+  protected def shift(i: Int): Int = index + i
+  protected def fetch(i: Int): A = element(shift(i))
+  protected def lowerMargin: Int = index
+  protected def upperMargin: Int = array.length - length - index
+
+  override def apply(i: Int): A =
+    if (i < 0 || i >= length) throw new ArrayIndexOutOfBoundsException(i)
+    else fetch(i)
+
+  def +:[B >: A](b: B): Spandex[B]
+
+  def :+[B >: A](b: B): Spandex[B]
+
+  override def ++[B >: A](xs: IterableOnce[B]): Spandex[B] = xs match {
+    case bs: Spandex[B] =>
+      Spandex.tabulate(length + bs.length) {
+        case i if i < length => apply(i)
+        case i => bs(i - length)
+      }
```

Comment:
You could probably have even better performance by using `System.arraycopy` here.